sky/skylet/job_lib.py

```diff
@@ -815,25 +830,28 @@ def cancel_jobs_encoded_results(jobs: Optional[List[int]],
         # it is being cancelled
         with filelock.FileLock(_get_lock_path(job_id)):
             job = _get_jobs_by_ids([job_id])[0]
-            if job['pid'] > 0:
+            if _is_job_driver_process_running(job['pid'], job_id):
                 # Not use process.terminate() as that will only terminate the
                 # process shell process, not the ray driver process
                 # under the shell.
-                # No need to use `subprocess_utils.kill_children_processes`,
-                # as the underlying ray driver process will handle the
-                # termination of user programs (we use SIGTERM to allow ray
-                # driver to gracefully exit).
-                try:
-                    os.killpg(job['pid'], signal.SIGTERM)
-                except ProcessLookupError:
-                    # The process may have already finished.
-                    pass
+                # We start a daemon to recursively kill the process group, and
+                # trigger the killing immediately by sending SIGTERM to the
+                # process group.
+                # This allows the job to be killed asynchronously.
+                subprocess_utils.kill_process_daemon(job['pid'])
+                os.killpg(job['pid'], signal.SIGTERM)
```

Comment:
Seems like there is a potential race here if the daemon can't get fetch the process children before we kill the process. Could be fixed if the daemon process waits to fork() after it's already got an initial list of the children.

Reply:
Good catch! I now pass the initial children process to the underlying daemon, so it should be able to have the children processes correctly even after the parent process is killed. : )